scripts/test_usb_scan.py

```diff
@@ -0,0 +1,393 @@
+#!/usr/bin/env python
+#
+# unittest for usb_scan.py
+
+from collections import Mapping, Container, Iterable
+import mock
+from nose.tools import nottest
+import os
+import shutil
+import sys
+import tempfile
+import unittest
+
+sys.modules["xcp"] = mock.Mock()
+sys.modules["xcp.logger"] = mock.Mock()
+sys.modules["pyudev"] = mock.Mock()
+
+log_file = None
+
+
+class MocDeviceAttrs(Mapping):
+    def __init__(self, device):
+        self.d = device.get_attr()
+
+    def __iter__(self):
+        for name in self.d:
+            yield name
+
+    def __len__(self):
+        return len(self.d)
+
+    def __getitem__(self, name):
+        return self.d.get(name)
+
+
+class MocDevice(Mapping):
+
+    def __init__(self, d):
+        self.d = d
+
+    @property
+    def sys_name(self):
+        return self.d.get("name")
+
+    def get_prop(self):
+        return self.d.get("props")
+
+    def get_attr(self):
+        return self.d.get("attrs")
+
+    @property
+    def attributes(self):
+        return MocDeviceAttrs(self)
+
+    def __iter__(self):
+        for name in self.get_prop():
+            yield name
+
+    def __len__(self):
+        return len(self.get_prop())
+
+    def __getitem__(self, name):
+        return self.get_prop().get(name)
+
+
+class MocEnumerator(object):
+
+    def __init__(self, ds):
+        self.ds = ds
+
+    def __iter__(self):
+        for d in self.ds:
+            yield MocDevice(d)
+
+
+class MocContext(object):
+
+    def __init__(self, devices, interfaces):
+        self.devices = devices
+        self.interfaces = interfaces
+
+    def list_devices(self, **kwargs):
+        if "usb" == kwargs.pop("subsystem"):
+            dev_type = kwargs.pop("DEVTYPE")
+            if "usb_device" == dev_type:
+                return MocEnumerator(self.devices)
+            elif "usb_interface" == dev_type:
+                return MocEnumerator(self.interfaces)
+        return MocEnumerator([])
+
+
+def mock_setup(mod, devices, interfaces, path):
+    mod.log.error = test_log
+    mod.log.debug = test_log
+    mod.Policy._PATH = path
+    mod.pyudev.Context = mock.Mock(return_value=MocContext(
+        devices, interfaces))
+
+
+@nottest
+def test_log(m):
```

Comment:
I would avoid using a file, instead I'd mock the stdout or the stderr and use print.
An example is
```python
>>> from StringIO import StringIO
>>> def foo():
...     print 'Something'
...
>>> @patch('sys.stdout', new_callable=StringIO)
... def test(mock_stdout):
...     foo()
...     assert mock_stdout.getvalue() == 'Something\n'
...
>>> test()
```
from the mock documentation at http://www.voidspace.org.uk/python/mock/patch.html

However it doesn't seem that you are using the logs content at all. In that case just replace the log funciton with a void function

Reply:
OK. remove log file, and make test_log empty.